File: C++/Card/UnitCard.cpp
//
// Created by DOEMsy on 2021/1/2.
//

#include "UnitCard.h"


#include <utility>

UnitCard::UnitCard() {
    this->Type = CardTypeMonster;
    this->SelfCombat = 0;
}

UnitCard::UnitCard(const std::string &Name, const std::string &Desc, const int Combat, const int Level,
                   const std::set<label> &Label) : Card(Name, Desc) {
    this->Type = CardTypeMonster;
    this->SelfCombat = Combat;
    this->Level = Level;
    this->Label = Label;
}

//单位卡出牌动作为在目标位置召唤一个属性相同的随从
int UnitCard::Play(const int cardType, Player *player, int Card_i, std::vector<int> to) {
    //类型不一样 不合法 拒绝出牌
    if (cardType != this->Type) return 0;
    if (to[0] > 0) {
        //登场
        CopyCard(player->HandCards, Card_i, player->Lines[to[0]]);
        //执行登场效果
        Debut();
    } else if (to[0] < 0) {
        //登场
        CopyCard(player->OpPlayer->HandCards, Card_i, player->OpPlayer->Lines[to[0]]);
        //执行登场效果
        Debut();
    } else {

    }
    //出牌成功
    return 1;
}

int UnitCard::Combat() {
    int res = SelfCombat;
    for (auto &se : Status) {
        res += se.CombatAmend();
    }
    return res;
}

File: Py/Card/UnitCard.py
from Card.Card import *


class UnitCard(Card):
    def __init__(self, name, desc, combat, level, label):
        super().__init__(name, desc)
        self.SelfCombat = combat
        self.Level = level
        self.Label = label
        self.Type = "UnitCard"
        self.Status = []

    def Play(self, player, to):
        if (3 >= to[0] > 0):
            player.Lines[to[0] - 1].append(self)
            self.Debut()
        elif (-3 <= to[0] < 0):
            player.OpPlayer.Lines[-1 - to[0]].append(self)
            self.Debut()
        else:
            return False
        return True

    def Combat(self):
        res = self.SelfCombat
        for se in self.Status:
            res += se.CombatAmend()
        return res

    def Debut(self):
        return True

    def OnCourt(self):
        return True

    def Round(self):
        return True

    def ToNextTurn(self):
        return True

    def Dead(self):
        return True

    def lstr(self):
        return "[{},{},{},{},{}]".format(self.Type, self.Name, self.Combat(), self.Level, self.Desc)

    def sstr(self):
        return "[{},{},{}]".format(self.Name, self.Combat(), self.Level)
